src/java.base/share/classes/jdk/internal/foreign/StringSupport.java

```diff
@@ -93,15 +115,23 @@ private static void writeFast_int(MemorySegment segment, long offset, Charset ch
         segment.set(JAVA_INT, offset + bytes.length, 0);
     }
 
-    private static int strlen_byte(MemorySegment segment, long start) {
-        // iterate until overflow (String can only hold a byte[], whose length can be expressed as an int)
-        for (int offset = 0; offset >= 0; offset++) {
-            byte curr = segment.get(JAVA_BYTE, start + offset);
-            if (curr == 0) {
-                return offset;
+    private static int native_strlen_byte(MemorySegment segment, long start) {
+        try {
+            if (start > 0) {
+                segment = segment.asSlice(start);
             }
+            long len = segment.byteSize() < MAX_TRIVIAL_SIZE
```

Comment:
When calling `strlen` should it be considered if segments is heap or off-heap?

Reply:
The linker can call native methods where the segment is heap-based too.

Reply:
Or, no. You are right. We have to handle this separately.